hyperspace/cosmos/src/provider.rs

```diff
@@ -0,0 +1,1149 @@
+	type Error = Error;
+
+	async fn query_latest_ibc_events<C>(
+		&mut self,
+		finality_event: Self::FinalityEvent,
+		counterparty: &C,
+	) -> Result<Vec<(Any, Vec<IbcEvent>, UpdateType)>, anyhow::Error>
+	where
+		C: Chain,
+	{
+		let finality_event_height = match finality_event {
+			FinalityEvent::Tendermint { from: _, to } => to,
+		};
+		let client_id = self.client_id();
+		let latest_cp_height = counterparty.latest_height_and_timestamp().await?.0;
+		let latest_cp_client_state =
+			counterparty.query_client_state(latest_cp_height, client_id.clone()).await?;
+		let client_state_response = latest_cp_client_state
+			.client_state
+			.ok_or_else(|| Error::Custom("counterparty returned empty client state".to_string()))?;
+		let client_state =
+			ClientState::<HostFunctionsManager>::decode_vec(&client_state_response.value)
+				.map_err(|_| Error::Custom("failed to decode client state response".to_string()))?;
+		let latest_cp_client_height = client_state.latest_height().revision_height;
+		let latest_height = self.latest_height_and_timestamp().await?.0;
+		let latest_revision = latest_height.revision_number;
+
+		let from = TmHeight::try_from(latest_cp_client_height).unwrap();
+		let to = finality_event_height;
+		log::info!(target: "hyperspace_cosmos", "Getting blocks {}..{}", from, to);
+
+		// query (exclusively) up to `to`, because the proof for the event at `to - 1` will be
+		// contained at `to` and will be fetched below by `msg_update_client_header`
+		let update_headers =
+			self.msg_update_client_header(from, to, client_state.latest_height).await?;
+		let mut block_events = Vec::new();
+		for height in from.value()..to.value() {
+			let ibc_events = self.parse_ibc_events_at(latest_revision, height).await?;
+			block_events.push(ibc_events);
+		}
+		// we don't submit events for the last block, because we don't have a proof for it
+		block_events.push(Vec::new());
+		assert_eq!(block_events.len(), update_headers.len(), "block events and updates must match");
+
+		let mut updates = Vec::new();
+		for (events, (update_header, update_type)) in block_events.into_iter().zip(update_headers) {
+			log::info!(target: "hyperspace_cosmos", "Fetching block {}", update_header.height().revision_height);
+			let update_client_header = {
+				let msg = MsgUpdateAnyClient::<LocalClientTypes> {
+					client_id: client_id.clone(),
+					client_message: AnyClientMessage::Tendermint(ClientMessage::Header(
+						update_header,
+					)),
+					signer: counterparty.account_id(),
+				};
+				let value = msg.encode_vec().map_err(|e| {
+					Error::from(format!("Failed to encode MsgUpdateClient {:?}: {:?}", msg, e))
+				})?;
+				Any { value, type_url: msg.type_url() }
+			};
+			updates.push((update_client_header, events, update_type));
+		}
+		Ok(updates)
+	}
+
+	// TODO: Changed result: `Item =` from `IbcEvent` to `IbcEventWithHeight` to include the
+	// necessary height field, as `height` is removed from `Attribute` from ibc-rs v0.22.0
```

Comment:
This comment does not match anymore.

Reply:
We still return `IbcEvent` here, so I think we eventually will replace it with `IbcEventWithHeight`

Reply:
I thought we could move away with just the `IbcEvent`. 
If so, makes sense.